File: lib/src/core/tag/tech/mifare/mifare.dart
part of core.tag;

/// The class provides access to `NFCMiFareTag` API for iOS.
class Mifare extends NFCTag {
  /// Constructs an instance with the given values for testing.
  ///
  /// The instances constructs by this way are not valid in
  /// the production environment.
  Mifare._({
    @required NFCTagPlatform delegate,
    this.mifareFamily,
    this.identifier,
    this.historicalBytes,
  }) : super._(delegate);

  /// Get an instance of `Mifare` for the given tag.
  ///
  /// Returns null if the tag is not compatible with `Mifare`.
  factory Mifare._from({
    @required NFCTagPlatform delegate,
  }) {
    if ( delegate.type.isNotMifare ) {
      return null;
    }

    final _data = Map<String, dynamic>.from(delegate.data);

    return Mifare._(
      delegate: delegate,
      identifier: Uint8List.fromList(
        _data['identifier'] as List<int>),
      mifareFamily: MifareFamily.values
          .firstWhere(
            (e) => e.value == _data['mifareFamily']),
      historicalBytes: Uint8List.fromList(
        _data['historicalBytes'] as List<int>));
  }

  /// The value from [NFCMiFareTag#mifareFamily] on iOS.
  final MifareFamily mifareFamily;

  /// The value from [NFCMiFareTag#identifier] on iOS.
  final Uint8List identifier;

  /// The value from [NFCMiFareTag#historicalBytes] on iOS.
  final Uint8List historicalBytes;

  /// Sends the native `Mifare` command to the tag.
  ///
  /// This uses [NFCMiFareTag#sendMiFareCommand] API on iOS.
  Future<Uint8List> sendCommand(Uint8List commandPacket) =>
      channel.invokeMethod(
        "Mifare", {
          'handle': handle,
          'method': 'sendCommand',
          'commandPacket': commandPacket,
        });

  /// Sends the `ISO7816 APDU` to the tag.
  ///
  /// This uses [NFCMiFareTag#sendMiFareISO7816Command] API on iOS.
  Future<Iso7816ResponseApdu> sendIso7816Command({
    int instructionClass,
    int instructionCode,
    int p1Parameter,
    int p2Parameter,
    Uint8List data,
    int expectedResponseLength,
  }) =>
      channel.invokeMethod<Map>(
        "Mifare", {
          'handle': handle,
          'method': 'sendIso7816Command',
          'instructionClass': instructionClass,
          'instructionCode': instructionCode,
          'p1Parameter': p1Parameter,
          'p2Parameter': p2Parameter,
          'data': data,
          'expectedResponseLength': expectedResponseLength,
        })
        .then(
          (value) => Iso7816ResponseApdu.fromMap(
            Map<String, dynamic>.from(value)));

  /// Sends the `ISO7816 APDU` to the tag.
  ///
  /// This uses [NFCMiFareTag#sendMiFareISO7816Command] API on iOS.
  Future<Iso7816ResponseApdu> sendIso7816CommandRaw(
      Uint8List data) =>
    channel.invokeMethod<Map>(
      "Mifare", {
        'handle': handle,
        'method': 'sendIso7816CommandRaw',
        'data': data,
      })
      .then(
        (value) => Iso7816ResponseApdu.fromMap(
          Map<String, dynamic>.from(value)));

  @override
  String toString() =>
      "[Mifare] "
      "identifier: ${identifier.toHexString()}, "
      "mifareFamily: $mifareFamily, "
      "historicalBytes: ${historicalBytes.toHexString()}";
}

File: ios/Classes/Mifare.swift
import CoreNFC
import Flutter

internal class Mifare: NSObject {
    private let tagManager: TagManager

    public init(_ tagManager: TagManager) {
        self.tagManager = tagManager
    }

    @available(iOS 13.0, *)
    public func handle(_ arguments: [String: Any?], result: @escaping FlutterResult) {
        switch arguments["method"] {
        case "sendCommand": sendCommand(arguments, result: result)
        case "sendIso7816Command": sendIso7816Command(arguments, result: result)
        case "sendIso7816CommandRaw": sendIso7816CommandRaw(arguments, result: result)
        default: result(FlutterMethodNotImplemented)
        }
    }

    @available(iOS 13.0, *)
    private func sendCommand(_ arguments: [String : Any?], result: @escaping FlutterResult) {
        self.tagManager.tagHandler(NFCMiFareTag.self, arguments, result) { tag in
            let commandPacket = (arguments["commandPacket"] as! FlutterStandardTypedData).data

            tag.sendMiFareCommand(commandPacket: commandPacket) { data, error in
                if let error = error {
                    result(getFlutterError(error))
                } else {
                    result(data)
                }
            }
        }
    }

    @available(iOS 13.0, *)
    private func sendIso7816Command(_ arguments: [String : Any?], result: @escaping FlutterResult) {
        self.tagManager.tagHandler(NFCMiFareTag.self, arguments, result) { tag in
            let apdu = NFCISO7816APDU(
                instructionClass: arguments["instructionClass"] as! UInt8,
                instructionCode: arguments["instructionCode"] as! UInt8,
                p1Parameter: arguments["p1Parameter"] as! UInt8,
                p2Parameter: arguments["p2Parameter"] as! UInt8,
                data: (arguments["data"] as! FlutterStandardTypedData).data,
                expectedResponseLength: arguments["expectedResponseLength"] as! Int)

            tag.sendMiFareISO7816Command(apdu) { payload, statusWord1, statusWord2, error in
                if let error = error {
                    result(getFlutterError(error))
                } else {
                    result([
                        "payload": payload,
                        "statusWord1": statusWord1,
                        "statusWord2": statusWord2,
                    ])
                }
            }
        }
    }

    @available(iOS 13.0, *)
    private func sendIso7816CommandRaw(_ arguments: [String : Any?], result: @escaping FlutterResult) {
        self.tagManager.tagHandler(NFCMiFareTag.self, arguments, result) { tag in
            guard let apdu = NFCISO7816APDU(data: (arguments["data"] as! FlutterStandardTypedData).data) else {
                result(
                    FlutterError(
                        code: "invalid_parameter",
                        message: nil,
                        details: nil))
                return
            }

            tag.sendMiFareISO7816Command(apdu) { payload, statusWord1, statusWord2, error in
                if let error = error {
                    result(getFlutterError(error))
                } else {
                    result([
                        "payload": payload,
                        "statusWord1": statusWord1,
                        "statusWord2": statusWord2,
                    ])
                }
            }
        }
    }
}